File: app/src/main/java/com/Fachhochschulebib/fhb/pruefungsplaner/Favoritenfragment.java
package com.Fachhochschulebib.fhb.pruefungsplaner;
//////////////////////////////
// favoritenfragment
//
//
// autor:
// inhalt:  stelllt die favorisierten prüfungen bereit.
// zugriffsdatum: 11.12.19
//
//
//
//////////////////////////////


import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CalendarView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.Fachhochschulebib.fhb.pruefungsplaner.data.AppDatabase;
import com.Fachhochschulebib.fhb.pruefungsplaner.data.TestPlanEntry;

import java.util.ArrayList;
import java.util.List;


public class Favoritenfragment extends Fragment {
    private RecyclerView recyclerView;
    private CalendarView calendar;
    private  Button btnSearch;
    MyAdapterfavorits mAdapter;
    List<Boolean> check = new ArrayList<>();

    // Datenbank initialisierung
    AppDatabase roomdaten = AppDatabase.getAppDatabase(getContext());

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        final View v = inflater.inflate(R.layout.terminefragment, container, false);

        //Komponenten  initialisieren für die Verwendung
        recyclerView = (RecyclerView) v.findViewById(R.id.recyclerView4);
        recyclerView.setHasFixedSize(true);
        //linear layout manager
        LinearLayoutManager layoutManager = new LinearLayoutManager(v.getContext());
        recyclerView.setLayoutManager(layoutManager);
        calendar = (CalendarView) v.findViewById(R.id.caCalender);
        btnSearch = (Button) v.findViewById(R.id.btnDatum);

        calendar.setVisibility(View.GONE);
        List<String> courses = new ArrayList<>();
        List<String> profnames = new ArrayList<>();
        List<String> dates = new ArrayList<>();
        List<String> examNo = new ArrayList<>();
        List<String> room = new ArrayList<>();
        List<String> form = new ArrayList<>();
        btnSearch.setVisibility(View.INVISIBLE);

        // Merlin Gürtler
        // Aktiviert den swipe listener
        enableSwipeToDelete();

        new Thread(new Runnable() {
            @Override
            public void run() {
                List<TestPlanEntry> ppeList = roomdaten.userDao().getFavorites(true);
                
                // Abfrage ob Prüfungen favorisiert wurden
                // Favorisierte Prüfungen für die Anzeige vorbereiten
                for (TestPlanEntry entry: ppeList) {
                    courses.add(entry.getModule() + " "
                            + entry.getCourse());
                    profnames.add(entry.getFirstExaminer() + " "
                            + entry.getSecondExaminer() + " "
                            + entry.getSemester());
                    dates.add(entry.getDate());
                    examNo.add(entry.getID());
                    room.add(entry.getRoom());
                    form.add(entry.getExamForm());
                    check.add(true);
                }

                // definiere adapter
                // übergabe der variablen an den Recyclerview Adapter, für die darstellung
                mAdapter = new MyAdapterfavorits(courses, profnames, dates, examNo,room, form);

                new Handler(Looper.getMainLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        recyclerView.setAdapter(mAdapter);
                    }
                });
            }
        }).start();

        recyclerView.addOnItemTouchListener(
                new RecyclerItemClickListener(getActivity(), new   RecyclerItemClickListener.OnItemClickListener() {
                    @Override
                    public void onItemClick( final View view, final  int position) {

                        //einblenden vom Textview für die weiteren Informationen
                        final TextView txtSecondScreen
                                = (TextView) view.findViewById(R.id.txtSecondscreen);
                        View viewItem
                                = recyclerView.getLayoutManager().findViewByPosition(position);
                        LinearLayout layoutinformationen
                                =(LinearLayout) viewItem.findViewById(R.id.linearLayout);


                        //überprüfung ob das linear layout geklickt wurde
                        layoutinformationen.setOnClickListener(new  View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                if (txtSecondScreen.getVisibility() == v.VISIBLE) {
                                    txtSecondScreen.setVisibility(v.GONE);
                                    check.set(position,false);

                                } else {

                                    // Start Merlin Gürtler
                                    for (int i = 0; i < recyclerView.getChildCount(); i++) {
                                        View holder = recyclerView.getLayoutManager().findViewByPosition(i);
                                        // Try and Catch, da die App crasht
                                        // wenn das Element nicht im View Port ist
                                        try {
                                            final TextView txtSecondScreen2
                                                    = (TextView) holder.findViewById(R.id.txtSecondscreen);

                                            if (txtSecondScreen2.getVisibility() == v.VISIBLE) {
                                                txtSecondScreen2.setVisibility(v.GONE);
                                            }
                                        }catch (Exception e) {
                                            Log.d("ERROR","NOT IN VIEW PORT " + e);
                                        }

                                    }
                                    // Ende Merlin Gürtler

                                    txtSecondScreen.setVisibility(v.VISIBLE);
                                    txtSecondScreen.setText(((MyAdapterfavorits) mAdapter)
                                                   .giveString(position));
                                }
                            }
                        });

                        /*
                        try{
                            if(check.get(position)) {
                                txtSecondScreen.setVisibility(v.VISIBLE);
                                txtSecondScreen.setText(((MyAdapterfavorits) mAdapter)
                                               .giveString(position));
                            }}
                        catch(Exception e){

                        }
                         */
                    }
                })
        );

        // Start Merlin Gürtler
        recyclerView.addOnChildAttachStateChangeListener(new RecyclerView.OnChildAttachStateChangeListener() {
            @Override
            public void onChildViewAttachedToWindow(View view) {
            }

            // Wenn ein Element den Viewport verlässt, wird
            // der zweite Screen zu geklappt
            @Override
            public void onChildViewDetachedFromWindow(View view) {
                final TextView txtSecondScreen
                        = (TextView) view.findViewById(R.id.txtSecondscreen);
                if (txtSecondScreen.getVisibility() == view.VISIBLE) {
                    txtSecondScreen.setVisibility(view.GONE);
                }
            }
        });
        // Ende Merlin Gürtler
        return v;
    }

    // Start Merlin Gürtler
    private void enableSwipeToDelete() {
        // try and catch, da es bei einer
        // Orientierungsänderung sonst zu
        // einer NullPointerException kommt
        try {
            // Definiert den Listener
            swipeListener swipeToDeleteCallback = new swipeListener(getContext(), true) {
                @Override
                public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int i) {
                    final int position = viewHolder.getAdapterPosition();
                    mAdapter.remove(position);
                }
            };

            // Setzt den Listener
            ItemTouchHelper itemTouchhelper = new ItemTouchHelper(swipeToDeleteCallback);
            itemTouchhelper.attachToRecyclerView(recyclerView);
        } catch (Exception e) {
            Log.d("Error", "Orientation error" + e);
        }
    }
    // Ende Merlin Gürtler
}

File: app/src/main/java/com/Fachhochschulebib/fhb/pruefungsplaner/Favoritenfragment.kt
package com.Fachhochschulebib.fhb.pruefungsplaner

import androidx.recyclerview.widget.RecyclerView
import com.Fachhochschulebib.fhb.pruefungsplaner.data.AppDatabase
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import android.widget.TextView
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.LinearLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView.OnChildAttachStateChangeListener
import androidx.recyclerview.widget.ItemTouchHelper
import java.lang.Exception
import java.util.ArrayList

import kotlinx.android.synthetic.main.terminefragment.*
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

//////////////////////////////
// favoritenfragment
//
//
// autor:
// inhalt:  stelllt die favorisierten prüfungen bereit.
// zugriffsdatum: 11.12.19
//
//
//
//////////////////////////////
/**
 * Class for the Fragment, where the user can see his selected exams.
 *
 * @author Alexander Lange (Email:alexander.lange@fh-bielefeld.de)
 * @since 1.6
 * @see Fragment
 */
class Favoritenfragment : Fragment() {
    private var mAdapter: RecyclerViewFavoritAdapter? = null
    private var check: MutableList<Boolean> = ArrayList()

    //TODO Alexander Lange Start
    private var filterChangeListenerPosition: Int? = null

    //TODO Alexander Lange End
    private lateinit var viewModel: MainViewModel

    /**
     * Overrides the onCreate()-Method, which is called first in the Fragment-LifeCycle.
     * In this Method, the global parameter which are independent of the UI get initialized,
     * like the App-SharedPreferences and the reference to the Room-Database
     *
     * @since 1.6
     *
     * @author Alexander Lange (E-Mail:alexander.lange@fh-bielefeld.de)
     *
     * @see Fragment.onCreate
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(
            requireActivity(),
            MainViewModelFactory(requireActivity().application)
        )[MainViewModel::class.java]
    }

    /**
     * Overrides the onViewCreated()-Method, which is called in the Fragment LifeCycle right after the onCreateView()-Method.
     *
     * @since 1.6
     * @author Alexander Lange (E-Mail:alexander.lange@fh-bielefeld.de)
     * @see Fragment.onViewCreated
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //From onCreate

        //From onCreateView
        //createAdapter()
        //Komponenten  initialisieren für die Verwendung
        initRecyclerview()
        currentPeriode.visibility = View.GONE
        // Merlin Gürtler
        // Aktiviert den swipe listener
        enableSwipeToDelete()

        // Ende Merlin Gürtler
        //TODO Alexander Lange Start
        //Filter.onFilterChangedListener.add { OnFilterChanged() }
        //filterChangeListenerPosition = Filter.onFilterChangedListener.size - 1
        //TODO Alexander Lange End
    }

    /**
     * Initializes the recyclerview, that shows the selected courses.
     *
     * @author Alexander Lange
     * @since 1.6
     */
    private fun initRecyclerview() {
        recyclerView4?.setHasFixedSize(true)
        //linear layout manager
        val layoutManager = LinearLayoutManager(view?.context)
        recyclerView4?.layoutManager = layoutManager
        viewModel.liveFilteredFavorits.observe(viewLifecycleOwner){
            recyclerView4.adapter = it?.let { it1 ->
                Filter.validateList(it1).toMutableList()
            }?.let { it2 -> RecyclerViewFavoritAdapter(it2,viewModel) }
        }
    }


    /**
     * Called, when the [MainActivity.Filter] of the [MainActivity]-Class changes.
     * It recreates the recyclerview, so the user can see the new filtered items.
     *
     * @author Alexander Lange
     * @since 1.6
     */
    private fun OnFilterChanged() {
        createAdapter()
    }

    /**
     * Creates the adapter for the recyclerview. Loads the chosen exams from the Room-Database
     * and passes them into an adapter for the recyclerview.
     *
     * @author Alexander Lange
     * @since 1.6
     */
    fun createAdapter() {
        val courses: MutableList<String> = ArrayList()
        val profnames: MutableList<String> = ArrayList()
        val dates: MutableList<String> = ArrayList()
        val examNo: MutableList<String> = ArrayList()
        val room: MutableList<String> = ArrayList()
        val form: MutableList<String> = ArrayList()
        val ppeList = viewModel.getFavorites(true)

        // Abfrage ob Prüfungen favorisiert wurden
        // Favorisierte Prüfungen für die Anzeige vorbereiten
        if (ppeList != null) {
            for (entry in ppeList) {
                if (!Filter.validateFilter(entry)) {
                    continue
                }
                courses.add(
                    entry.module + " "
                            + entry?.course
                )
                profnames.add(
                    entry.firstExaminer + " "
                            + entry.secondExaminer + " "
                            + entry.semester
                )
                dates.add(entry.date ?: "")
                examNo.add(entry.id ?: "")
                room.add(entry.room ?: "")
                form.add(entry.examForm ?: "")
                check.add(true)
            }
        }
        // definiere adapter
        // übergabe der variablen an den Recyclerview Adapter, für die darstellung
    }

    //TODO Alexander Lange Start
    /**
     * Removes the listener, when this fragment is not longer visible.
     *
     * @author Alexander Lange
     * @since 1.6
     * @see Fragment.onDestroy
     * @see MainActivity.Filter
     */
    override fun onDestroy() {
        super.onDestroy()
        if (filterChangeListenerPosition != null) {
           // Filter.onFilterChangedListener.removeAt(filterChangeListenerPosition!!)
        }
    }
    //TODO Alexander Lange End

    /**
     * Overrides the onCreateView()-Method. It sets the current view to the terminefragment-layout.
     *
     * @return Returns the initialized view of this Fragment
     * @since 1.6
     * @author Alexander Lange (E-Mail:alexander.lange@fh-bielefeld.de)
     * @see Fragment.onCreateView
     */
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val v = inflater.inflate(R.layout.terminefragment, container, false)
        return v
    }

    /**
     * Creates the swipe-gesture on the recyclerview,
     * so the user can swipe to delete an entry from the list.
     *
     * @author Alexander Lange
     * @since 1.6
     */
    // Start Merlin Gürtler
    private fun enableSwipeToDelete() {
        // try and catch, da es bei einer
        // Orientierungsänderung sonst zu
        // einer NullPointerException kommt
        try {
            // Definiert den Listener
            val swipeToDeleteCallback: swipeListener = object : swipeListener(context!!, true) {
                override fun onSwiped(viewHolder: RecyclerView.ViewHolder, i: Int) {
                    val position = viewHolder.adapterPosition
                    mAdapter?.remove(position)
                }
            }

            // Setzt den Listener
            val itemTouchhelper = ItemTouchHelper(swipeToDeleteCallback)
            itemTouchhelper.attachToRecyclerView(recyclerView4)
        } catch (e: Exception) {
            Log.d("Error", "Orientation error$e")
        }
    } // Ende Merlin Gürtler
}